src/java.base/share/classes/sun/util/locale/LanguageTag.java

```diff
@@ -414,6 +407,50 @@ private boolean parsePrivateuse(StringTokenIterator itr, ParseStatus sts) {
         return found;
     }
 
+    public static String caseFoldTag(String tag) {
+        LanguageTag parsedTag = parse(tag, null);
+        // Illegal tags
+        if (parsedTag.parseSts.errorMsg != null) {
+            throw new IllformedLocaleException(String.format("Ill formed tag:" +
+                    " %s", parsedTag.parseSts.errorMsg));
+        }
+        // Legacy tags
+        String potentialLegacy = tag.toLowerCase(Locale.ROOT);
+        if (LEGACY.containsKey(potentialLegacy)) {
+            return LEGACY.get(potentialLegacy)[0];
+        }
+        // Non-legacy tags
+        StringBuilder bldr = new StringBuilder();
+        String[] subtags = tag.split("-");
+        boolean privateFound = false;
+        boolean singletonFound = false;
+        boolean privUseVarFound = false;
+        for (int i = 0; i < subtags.length; i++) {
+            if (privUseVarFound) {
+                bldr.append(subtags[i]);
+            } else if (i > 0 && isVariant(subtags[i]) && !singletonFound && !privateFound) {
+                bldr.append(subtags[i]);
+            } else if (i > 0 && isRegion(subtags[i]) && !singletonFound && !privateFound) {
+                bldr.append(canonicalizeRegion(subtags[i]));
+            } else if (i > 0 && isScript(subtags[i]) && !singletonFound && !privateFound) {
+                bldr.append(canonicalizeScript(subtags[i]));
+            // If subtag is not 2 letter, 4 letter, or variant
+            // under the right conditions, then it should be lower-case
+            } else {
+                if (isPrivateusePrefix(subtags[i])) {
+                    privateFound = true;
+                } else if (isExtensionSingleton(subtags[i])) {
+                    singletonFound = true;
+                } else if (subtags[i].equals(PRIVUSE_VARIANT_PREFIX)) {
+                    privUseVarFound = true;
+                }
```

Comment:
These flags can become true but inside the loop cannot become false again.
Is that correct?  For example, I think there can be multiple extension singletons.

Reply:
That was intentional as the booleans represent if a private prefix, singleton prefix, or "lvariant" has been found in general (not that the current position is the suffix of one of those prefixes).

Consider if there is a multiple extension singleton (eg: `a-BBB-CCC-DDD`)

Once singletonFound is set to `true`, it is assumed all further subtags should be lowercased unless it is a private use prefixed by "lvariant". Thus it would be folded to `a-bbb-ccc-ddd` as expected.

Did that address your point?